content/blog/esp32-bootstrapping/index.md

```diff
@@ -0,0 +1,375 @@
+
+The main building blocks of the article we will be investigating are:
+
+- **ROM code & loader** *(image loader, reserved memory regions)*
+- **2nd stage loader** *(build types, bootstrap flow)*
+- **Building** *(linking, memory utilization)*
+- **Tooling** *(esptool.py, image formats)*
+
+## ROM code
+
+The ESP32 SoCs has external flash memory which need certain attention prior to use. That is one of the reason why ESP32 has on-chip ROM memory. All necessary startup, communication and the flash access libraries are stored and directly accessed from there.
+
+There are certain functions available in the ROM code that are particularly useful (debugging, loading purposes) and can have an impact on the overall code footprint. However, using or not using the ROM functions during the application runtime is completely up to the user.
+
+### ROM loader
+
+Main feature of ROM code is to load the image present at the flash non-volatile memory. Therefore we can refer to it as the *1st stage bootloader*.
+
+The binary image loaded by the ROM code must take a form recognizable by the loader. It uses the ESP image header and ESP segment header to identify the image and its subsequent segments.
+
+The ***ESP image format*** header (24 Bytes):
+
+```cpp
+typedef struct {
+  uint8_t magic;              /* Magic word 0xE9 */
+  uint8_t segment_count;      /* Count of memory segments */
+  uint8_t spi_mode;           /* flash read mode */
+  uint8_t spi_speed: 4;       /* flash frequency */
+  uint8_t spi_size: 4;        /* flash chip size */
+  uint32_t entry_addr;        /* Entry address */
+  uint8_t wp_pin;             /* WP pin when SPI pins set via efuse
+                               * 0xEE=disabled
+			       */
+  uint8_t spi_pin_drv[3];     /* Drive settings for the SPI flash pins */
+  uint16_t chip_id;           /* Chip identification number */
+  uint8_t min_chip_rev;       /* Minimal chip revision supported by image */
+  uint16_t min_chip_rev_full; /*   format: major * 100 + minor */
+  uint16_t max_chip_rev_full; /*   format: major * 100 + minor */
+  uint8_t reserved[4];        /* Reserved bytes in additional header space */
+  uint8_t hash_appended;      /* If 1, a SHA256 digest "simple hash"
+                               * is appended after the checksum.
+                               */
+} esp_image_header_t;
+
+```
+
+The ***ESP image format*** segment header (8 Bytes):
+```cpp
+typedef struct {
+    uint32_t load_addr;     /* Address of segment */
+    uint32_t data_len;      /* Length of data */
+} esp_image_segment_header_t;
+
+```
+
+Graphical representation of the ESP image format:
+
+{{< figure
+    default=true
+    src="img/esp-image-format.webp"
+    alt=""
+    caption="The ESP image format."
+    >}}
+
+This binary image format is used in 1st stage image loading, but can be adopted and used for other purposes as well. For example when providing an additional images for asymmetric multi processing (AMP).
```

Comment:
an additional ~~images~~ image

Reply:
this one is actually correct, there can be more than one image (2xcpu 1xlpcpu)